gateway/loaders/zookeeper/zookeeper.go

```diff
@@ -0,0 +1,343 @@
+			},
+		}
+	}
+	return configs, nil
+}
+
+func (z *ZookeeperTargetLoader) GetConfiguration() (*targetpb.Configuration, error) {
+	zClient, err := NewZookeeperClient(z.hosts, z.config.ZookeeperTimeout)
+	if err != nil {
+		return nil, err
+	}
+	data, err := zClient.GetZookeeperData()
+	switch err {
+	case zk.ErrNoNode:
+		configs := &targetpb.Configuration{}
+		z.config.Log.Info().Msg("Could not parse zookeeper configuration: " + err.Error())
+		if exists, _, _ := zClient.conn.Exists(CleanPath(TargetPath)); !exists {
+			z.config.Log.Info().Msg("Creating path: " + TargetPath)
+			if err := CreateFullPath(zClient.conn, TargetPath, zk.WorldACL(zk.PermAll)); err != nil {
+				return nil, err
+			}
+		}
+		if exists, _, _ := zClient.conn.Exists(CleanPath(RequestPath)); !exists {
+			z.config.Log.Info().Msg("Creating path: " + RequestPath)
+			if err := CreateFullPath(zClient.conn, RequestPath, zk.WorldACL(zk.PermAll)); err != nil {
+				return nil, err
+			}
+		}
+		return configs, nil
+	case nil:
+		z.config.Log.Info().Msg("Succesfully retreived zookeeper configuration")
+	default:
+		return nil, err
+	}
+
+	configs, err := z.zookeeperToTargets(data)
+
+	if err := target.Validate(configs); err != nil {
+		return nil, fmt.Errorf("configuration in zookeeper is invalid: %v", err)
+	}
+
+	return configs, nil
+}
+
+func (z *ZookeeperTargetLoader) Start() error {
+	_, err := z.GetConfiguration() // make sure there are no errors at startup
+	return err
+}
+
+func (z *ZookeeperTargetLoader) WatchConfiguration(targetChan chan<- *connections.TargetConnectionControl) error {
+	for {
+		targetConfig, err := z.GetConfiguration()
+		if err != nil {
+			z.config.Log.Error().Err(err).Msgf("Unable to get target configuration.")
+		} else {
+			controlMsg := new(connections.TargetConnectionControl)
+			if z.last != nil {
+				for targetName := range z.last.Target {
+					_, exists := targetConfig.Target[targetName]
+					if !exists {
+						controlMsg.Remove = append(controlMsg.Remove, targetName)
+					}
+				}
+			}
+			controlMsg.Insert = targetConfig
+			z.last = targetConfig
+
+			targetChan <- controlMsg
+		}
+		time.Sleep(z.interval)
+	}
+}
+
+func (z *ZookeeperClient) getNodeValue(nodePath string) ([]byte, error) {
+	data, _, err := z.conn.Get(nodePath)
+	if err != nil {
+		log.Log.Error(err, "zNode get failed", "path", nodePath)
+		return nil, err
+	}
+	return data, nil
+}
+
+func CleanPath(path string) string {
+	return "/" + strings.Trim(path, "/")
+}
+
+func (z *ZookeeperClient) ConnectToZookeeper(hosts []string, timeout time.Duration) (*zk.Conn, error) {
+	newConn, eventChan, err := zk.Connect(hosts, timeout)
+	if err != nil {
+		return nil, err
+	}
+	go z.zookeeperEventHandler(eventChan)
+	return newConn, nil
+}
+
+func (z *ZookeeperClient) zookeeperEventHandler(zkEventChan <-chan zk.Event) {
+	var disconnectedCount int
+	for event := range zkEventChan {
+		switch event.State {
+		case zk.StateDisconnected:
+			disconnectedCount++
```

Comment:
The disconnect count isn't used anywhere. Might be useful for debug purposes though, let's log it. We might have something like:

```
case zk.StateDisconnected:
			z.config.Log.Info().Msg("Disconnected. Reconnect attempts:  " + disconnectCount)
			disconnectedCount++
case zk.StateHasSession:
			z.config.Log.Info().Msg("Reconnected.")
			disconnectedCount = 0
```